SourceGen MauiXamlInflator using wrong type when working with Enum and extension class

### Description

Having an Enum and a static extension class for this enum. When now trying to use that Enum in Xaml and the name of the extension class is [NameOfEnum]Extension, SourceGen will generate code that references the extension class accidentally instead of the Enum itself, which won't compile.

### Steps to Reproduce

1. Create new MAUI project from template and add a file with an enum definition + a static extension class for it.

```
    public enum MyEnum
    {
        Value1,
        Value2,
        Value3
    }

    // Rename to any name except [NameOfEnum]Extension and the issue is resolved.
    internal static class MyEnumExtension
    {
        public static string ToFriendlyString(this MyEnum value)
        {
        }
    }
```

2. Create a random typed class with a property that accepts enums

```
    internal class MyDataObject<T> where T: Enum
    {
        public T? EnumValue { get; set; }
    }
```

3. Use the `MyDataObject` with `MyEnum` somewhere in Xaml, e.g. as a resource:

```
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:enums="clr-namespace:MauiXamlInflatorBug.Enums"
             xmlns:app="clr-namespace:MauiXamlInflatorBug"
             x:Class="MauiXamlInflatorBug.MainPage">

    <ContentPage.Resources>
        <ResourceDictionary>
            <app:MyDataObject x:Key="MyDataObject" EnumValue="{x:Static enums:MyEnum.Value1}" x:TypeArguments="enums:MyEnum" />
        </ResourceDictionary>
    </ContentPage.Resources>
```

4. Ensure `MauiXamlInflator` is set to `SourceGen` in the project file
5. **The project won't build.** Check the generated file for `MainPage` and see that it uses the name of the extension class there instead of the enum itself

```
    var myDataObject = new global::MauiXamlInflatorBug.MyDataObject<global::MauiXamlInflatorBug.Enums.MyEnumExtension>();
    ...
    myDataObject.EnumValue = (global::MauiXamlInflatorBug.Enums.MyEnumExtension)object0;
```

6. Rename the extension class `MyEnumExtension` to something else, e.g. `MyCoolEnumExtension` or even `MyEnumExtensions`
7. Build the project again. Now it succeeds. Check the generated code and see that it now uses the Enum correctly

```
    var myDataObject = new global::MauiXamlInflatorBug.MyDataObject<global::MauiXamlInflatorBug.Enums.MyEnum>();
    ....
    myDataObject.EnumValue = (global::MauiXamlInflatorBug.Enums.MyEnum)staticExtension;
```

### Link to public reproduction project repository

https://github.com/tschramme86/MauiXamlInflatorBug

### Version with bug

10.0.30

### Is this a regression from previous behavior?

No, this is something new, Not sure, did not test other versions

### Last version that worked well

_No response_

### Affected platforms

I was *not* able test on other platforms

### Affected platform versions

Android

### Did you find any workaround?

Renaming the extension class as described above.

### Relevant log output

```shell
Compiler output for the case described above:


1>CSC : error MAUIG1001: An error occured while parsing Xaml: Member not found.
1>D:\Source\MauiXamlInflatorBug\obj\Debug\net10.0-android\Microsoft.Maui.Controls.SourceGen\Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator\MainPage.xaml.xsg.cs(21,67,21,116): error CS0718: 'MyEnumExtension': static types cannot be used as type arguments
1>D:\Source\MauiXamlInflatorBug\obj\Debug\net10.0-android\Microsoft.Maui.Controls.SourceGen\Microsoft.Maui.Controls.SourceGen.CodeBehindGenerator\MainPage.xaml(10,28,10,86): error CS0716: Cannot convert to static type 'MyEnumExtension'
```

--- a/src/Controls/src/Build.Tasks/XmlTypeExtensions.cs
+++ b/src/Controls/src/Build.Tasks/XmlTypeExtensions.cs
@@ -111,7 +111,14 @@ namespace Microsoft.Maui.Controls.Build.Tasks
 				string typeName = typeInfo.typeName.Replace('+', '/'); //Nested types
 				var type = module.GetTypeDefinition(cache, (typeInfo.assemblyName, typeInfo.clrNamespace, typeName));
 				if (type is not null && type.IsPublicOrVisibleInternal(module))
+				{
+					// Skip static classes found via the "Extension" suffix expansion when they shadow
+					// the exact name (e.g. MyEnumExtension shadowing MyEnum). Static types cannot
+					// implement IMarkupExtension, so they are never valid markup extensions. (#34021)
+					if (typeInfo.typeName != xmlType.Name && type.IsAbstract && type.IsSealed)
+						return null;
 					return type;
+				}
 				return null;
 			}, expandToExtension: expandToExtension);
 

--- a/src/Controls/src/SourceGen/XmlTypeExtensions.cs
+++ b/src/Controls/src/SourceGen/XmlTypeExtensions.cs
@@ -67,6 +67,12 @@ static class XmlTypeExtensions
 			if (types.Length == 1)
 			{
 				symbol = types[0];
+				// If the match is via the "Extension" suffix but the type is static, it cannot be
+				// an IMarkupExtension (static types cannot implement interfaces). In that case, skip
+				// this match and continue to try the exact name. This prevents a static helper class
+				// like MyEnumExtension from being returned instead of the MyEnum enum type. (#34021)
+				if (suffix == "Extension" && symbol.IsStatic)
+					continue;
 				if (symbol.IsGenericType && xmlType.TypeArguments is not null)
 				{
 					var typeArgs = xmlType.TypeArguments.Select(typeArg => typeArg.GetTypeSymbol(reportDiagnostic, compilation, xmlnsCache, typeCache)!).ToArray();

--- a/src/Controls/src/Xaml/XmlTypeXamlExtensions.cs
+++ b/src/Controls/src/Xaml/XmlTypeXamlExtensions.cs
@@ -108,6 +108,15 @@ namespace Microsoft.Maui.Controls.Xaml
 				if ((returnTypeName == null || returnTypeName == typeInfo.typeName) //only return multiple types if they share the same name. avoid returning both BindingExtension and Binding
 					&& (type = refFromTypeInfo(typeInfo)) != null)
 				{
+					// If this is an Extension-suffix match but the type is static (cannot implement
+					// IMarkupExtension), skip it and continue to try the exact name. This prevents
+					// a static helper class like MyEnumExtension from shadowing MyEnum. (#34021)
+					if (returnTypeName == null
+						&& typeInfo.typeName.EndsWith("Extension", StringComparison.Ordinal)
+						&& typeInfo.typeName != elementName
+						&& type is System.Type runtimeType
+						&& runtimeType.IsAbstract && runtimeType.IsSealed) // static in reflection: abstract+sealed
+						continue;
 					returnTypeName = typeInfo.typeName;
 					yield return type;
 				}

--- a/src/Controls/tests/Xaml.UnitTests/Issues/Maui34021.xaml
+++ b/src/Controls/tests/Xaml.UnitTests/Issues/Maui34021.xaml
@@ -2,7 +2,7 @@
 <ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
              xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
              xmlns:local="clr-namespace:Microsoft.Maui.Controls.Xaml.UnitTests"
-             x:Class="Microsoft.Maui.Controls.Xaml.UnitTests.Maui34021Rtxc">
+             x:Class="Microsoft.Maui.Controls.Xaml.UnitTests.Maui34021">
 
-     <local:Maui34021DataObject x:TypeArguments="local:Maui34021MyEnum"/>
+        <local:Maui34021DataObject x:TypeArguments="local:Maui34021MyEnum"/>
 </ContentPage>

--- a/src/Controls/tests/Xaml.UnitTests/Issues/Maui34021.xaml
+++ b/src/Controls/tests/Xaml.UnitTests/Issues/Maui34021.xaml
@@ -1,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             xmlns:local="clr-namespace:Microsoft.Maui.Controls.Xaml.UnitTests"
+             x:Class="Microsoft.Maui.Controls.Xaml.UnitTests.Maui34021Rtxc">
+
+     <local:Maui34021DataObject x:TypeArguments="local:Maui34021MyEnum"/>
+</ContentPage>

--- a/src/Controls/tests/Xaml.UnitTests/Issues/Maui34021.xaml.cs
+++ b/src/Controls/tests/Xaml.UnitTests/Issues/Maui34021.xaml.cs
@@ -1,0 +1,83 @@
+using System;
+using Xunit;
+
+using static Microsoft.Maui.Controls.Xaml.UnitTests.MockSourceGenerator;
+
+namespace Microsoft.Maui.Controls.Xaml.UnitTests;
+
+public enum Maui34021MyEnum { Value1, Value2, Value3 }
+
+internal static class Maui34021MyEnumExtension
+{
+    public static string ToFriendlyString(this Maui34021MyEnum value) => value.ToString();
+}
+
+public class Maui34021DataObject<T> : View where T : Enum
+{
+}
+
+public partial class Maui34021Rtxc : ContentPage
+{
+    public Maui34021Rtxc() => InitializeComponent();
+
+    [Collection("Issue")]
+    public class Tests : IDisposable
+    {
+        public void Dispose() => Application.Current = null;
+
+        [Theory]
+        [InlineData(XamlInflator.Runtime)]
+        [InlineData(XamlInflator.XamlC)]
+        [InlineData(XamlInflator.SourceGen)]
+        internal void SourceGenResolvesEnumTypeNotExtensionClass(XamlInflator inflator)
+        {
+            if (inflator == XamlInflator.SourceGen)
+            {
+                const string xaml = """
+<?xml version="1.0" encoding="utf-8" ?>
+<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             xmlns:local="clr-namespace:Microsoft.Maui.Controls.Xaml.UnitTests"
+             x:Class="Microsoft.Maui.Controls.Xaml.UnitTests.Maui34021SourceGenRepro">
+    <local:Maui34021DataObject x:TypeArguments="local:Maui34021MyEnum" />
+</ContentPage>
+""";
+
+                const string csharp = """
+using System;
+using Microsoft.Maui.Controls;
+
+namespace Microsoft.Maui.Controls.Xaml.UnitTests;
+
+public enum Maui34021MyEnum { Value1, Value2, Value3 }
+
+internal static class Maui34021MyEnumExtension
+{
+    public static string ToFriendlyString(this Maui34021MyEnum value) => value.ToString();
+}
+
+public class Maui34021DataObject<T> : View where T : Enum
+{
+}
+
+public partial class Maui34021SourceGenRepro : ContentPage
+{
+    public Maui34021SourceGenRepro() => InitializeComponent();
+}
+""";
+
+                var result = CreateMauiCompilation()
+                .WithAdditionalSource(csharp)
+                .RunMauiSourceGenerator(new AdditionalXamlFile("Issues/Maui34021SourceGenRepro.xaml", xaml, TargetFramework: "net10.0"));
+                var generated = result.GeneratedInitializeComponent();
+                Assert.DoesNotContain("Maui34021MyEnumExtension", generated, StringComparison.Ordinal);
+                Assert.Contains("Maui34021MyEnum", generated, StringComparison.Ordinal);
+            }
+            else
+            {
+                var page = new Maui34021Rtxc(inflator);
+                Assert.IsType<Maui34021DataObject<Maui34021MyEnum>>(page.Content);
+            }
+        }
+    }
+}

--- a/src/Controls/tests/Xaml.UnitTests/Issues/Maui34021.xaml.cs
+++ b/src/Controls/tests/Xaml.UnitTests/Issues/Maui34021.xaml.cs
@@ -16,9 +16,9 @@ public class Maui34021DataObject<T> : View where T : Enum
 {
 }
 
-public partial class Maui34021Rtxc : ContentPage
+public partial class Maui34021 : ContentPage
 {
-    public Maui34021Rtxc() => InitializeComponent();
+    public Maui34021() => InitializeComponent();
 
     [Collection("Issue")]
     public class Tests : IDisposable
@@ -75,7 +75,7 @@ public partial class Maui34021SourceGenRepro : ContentPage
             }
             else
             {
-                var page = new Maui34021Rtxc(inflator);
+                var page = new Maui34021(inflator);
                 Assert.IsType<Maui34021DataObject<Maui34021MyEnum>>(page.Content);
             }
         }

